File: 2020/day23.c
#include <stdio.h>

#define MAX_VALUE 1000000
#define MAX_INITIAL 9

struct {
    int nodes[MAX_VALUE + 1];
    int max_value;
    int current;
} ring;

void init(const char *initial, int max_value)
{
    ring.max_value = max_value;
    int prev = max_value <= MAX_INITIAL ? initial[MAX_INITIAL-1] : max_value;
    for (int i = 0; i < MAX_INITIAL; i++) {
        int n = initial[i] - '0';
        ring.nodes[prev] = n;
        prev = n;
    }
    for (int n = MAX_INITIAL + 1; n <= MAX_VALUE; n++) {
        ring.nodes[prev] = n;
        prev = n;
    }
    ring.current = initial[0] - '0';
}

void move(int times)
{
    for (int i = 0; i < times; i++) {
        int n1 = ring.nodes[ring.current];
        int n2 = ring.nodes[n1];
        int n3 = ring.nodes[n2];
        int next_value = ring.current;
        while (1) {
            next_value = next_value == 1 ? ring.max_value : next_value - 1;
            if (next_value != n1 && next_value != n2 && next_value != n3) {
                break;
            }
        }
        ring.nodes[ring.current] = ring.nodes[n3];
        ring.nodes[n3] = ring.nodes[next_value];
        ring.nodes[next_value] = n1;
        ring.current = ring.nodes[ring.current];
    }
}

int solution1(void)
{
    int n = ring.nodes[1];
    int result = 0;
    while (n != 1) {
        result = result * 10 + n;
        n = ring.nodes[n];
    }
    return result;
}

long long solution2(void)
{
    return (long long) ring.nodes[1] * ring.nodes[ring.nodes[1]];
}

int main()
{
    init("871369452", MAX_VALUE);
    move(10000000);
    printf("%lld\n", solution2());
    return 0;
}

File: 2020/day23.py
import adventofcode

class node_ring:
    def __init__(self, first_numbers, max_value):
        self.max_value = max_value
        self.nodes = [0] * (max_value + 1)

        prev = max_value
        if max_value <= len(first_numbers):
            prev = first_numbers[-1]
        for n in first_numbers:
            self.nodes[prev] = n
            prev = n
        for n in range(len(first_numbers) + 1, max_value + 1):
            self.nodes[prev] = n
            prev = n
        self.current = first_numbers[0]

    def move(self, times = 1):
        for x in range(times):
            n1 = self.nodes[self.current]
            n2 = self.nodes[n1]
            n3 = self.nodes[n2]
            next_value = self.current
            while True:
                next_value -= 1
                if next_value == 0:
                    next_value = self.max_value
                if next_value != n1 and next_value != n2 and next_value != n3:
                    break
            self.nodes[self.current] = self.nodes[n3]
            self.nodes[n3] = self.nodes[next_value]
            self.nodes[next_value] = n1
            self.current = self.nodes[self.current]

    def print(self):
        start_value = self.current
        print(start_value, end = '')
        n = start_value
        while self.nodes[n] != start_value:
            n = self.nodes[n]
            print('-', end = '')
            print(n, end = '')
        print()

    def solution1(self):
        n = self.nodes[1]
        result = []
        while n != 1:
            result.append(str(n))
            n = self.nodes[n]
        return ''.join(result)

    def solution2(self):
        print(self.nodes[1], self.nodes[self.nodes[1]])
        return self.nodes[1] * self.nodes[self.nodes[1]]

class runner(adventofcode.runner):
    def __init__(self):
        super().__init__(23)

    def reset(self):
        self.input = []

    def input_line(self, line):
        self.input = list(map(int, line))

    def solve1(self):
        ring = node_ring(self.input, 9)
        ring.move(100)
        return str(ring.solution1())

    def solve2(self):
        ring = node_ring(self.input, 1000000)
        ring.move(10000000)
        return str(ring.solution2())

r = runner()

r.test('Sample 1', ['389125467'], '67384529')

r.run()
